### PR TITLE
3.7 preparation

### LINKED ISSUE #24: Redesign client lifecycle
As of 3.6, Cassandra clients can be shared or non shared. Before querying, the user must invoke `connect`. There is no method for the user to _close_ a client, it can only be _disconnected_.

This lifecycle design is not aligned with the rest of Vert.x stack and does not fit well the typical Vert.x application model (with different verticles, possibly multiple instances).

Most Vert.x data clients work as follows:

- client connects automatically on first request
- client disconnects automatically when calling `close` (immediately if non shared or later if other shared clients are still running)
- `close` method is idempotent (user may invoke it any time in verticle code and yet undeploying the verticle later will not eagerly disconnect another shared client still running)

If the client connects automatically, the default keyspace must be configurable in the client options.

### LINKED ISSUE #25: Row stream end method invoked before all rows are emitted
To observe the problem, create a stream for a statement with fetch size bigger than the number of results. After a few rows have been emitted, pause the stream, then resume some time later. The stream emits a few more rows and then invokes the end handler before the end.

### LINKED ISSUE #26: Simplify and accelerate test suite
The test suite uses Vert.x Unit and Cassandra Unit. This is fine but we could leverage more features from both tools.

- Vert.x Unit `TestContext` object has methods to deal with asynchronous results (e.g. `asyncAssertSuccess`) but we don't use them. Instead, we often use Vert.x `Future` with composition and validate the result manually.
- Cassandra Unit has data loader classes to prepare a keyspace and run queries. They can be used instead of using the Vert.x client itself.[1]

Also, Cassandra clients instances are not managed with JUnit lifecycle methods.

Eventually, the base test class initializes two keyspaces before each test, and one of them is populated with a huge number of rows. These keyspaces could be created on demand and the number or rows inserted configured as needed.

[1] During the project inception, using our own dog food was important, as it was our main source of design feedback. But now test setup efficiency is more important.

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #24 , fixes #25 , fixes #26

### CODE REVIEW

File: src/test/java/io/vertx/cassandra/ExecutionTest.java [hunk 1]
```diff
@@ -1,169 +1,73 @@
...
-      return queryResult;
-    }).compose((List<Row> resultSet) -> {
-      for (Row row: resultSet) {
+  public void simpleExecuteWithBigAmountOfFetches(TestContext testContext) {
+    initializeRandomStringKeyspace(50);
+    String query = "select random_string from random_strings.random_string_by_first_letter where first_letter = 'B'";
+    SimpleStatement statement = new SimpleStatement(query);
+    // we would like to test that we are able to handle a lot of fetches.
+    // that is why we are setting fetch size here to 1
+    statement.setFetchSize(3);
```
Comment (Sammers21) [inline src/test/java/io/vertx/cassandra/ExecutionTest.java]: The number _3_ and comment _we are setting fetch size here to 1_ are inconsistent.

**Comment by Sammers21** [comment]
The build now is 2 times faster on my computer, just 20 seconds! Good job.

**Comment by tsegismont** [comment]
>
> The build now is 2 times faster on my computer, just 20 seconds! Good job.
>

Thank you!

File: src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java [hunk 2]
```diff
@@ -231,36 +124,88 @@ public CassandraClient queryStream(String sql, Handler<AsyncResult<CassandraRowS
...
+    if (current.session != null) {
+      future.complete(current.session);
+      return;
+    }
+    Cluster.Builder builder = options.dataStaxClusterBuilder();
+    if (builder.getContactPoints().isEmpty()) {
+      builder.addContactPoint(CassandraClientOptions.DEFAULT_HOST);
+    }
+    Cluster cluster = builder.build();
+    Session session = cluster.connect(options.getKeyspace());
```
[the remark this answers is not in the document]
Comment (Sammers21) [reply src/main/java/io/vertx/cassandra/impl/CassandraClientImpl.java]: Good

File: src/test/java/io/vertx/cassandra/ExecutionTest.java (hunk 1 above)
Comment (Sammers21) [reply src/test/java/io/vertx/cassandra/ExecutionTest.java]: I think this is not resolved since we still have inconsistency.

Comment (tsegismont) [reply src/test/java/io/vertx/cassandra/ExecutionTest.java]: I had forgotten to push the last commit

**Pull request merged**